pages/docs/couple-your-code/couple-your-code-waveform.md

```diff
@@ -0,0 +1,111 @@
+
+If we want to improve the accuracy by using waveforms, this requires an extension of the existing API, because we need a way to tell preCICE where we want to sample the waveform. For this purpose, preCICE offers an experimental API, which is currently only supporting a single linear interpolation along the complete time window. Here, `readBlockVectorData` accepts an additional argument `relativeReadTime`. This allows us to choose the time where the interpolant should be sampled:
+
+```cpp
+// stable API with constant data in time window
+void readBlockVectorData(int dataID, int size, const int* valueIndices, double* values) const;
+
+// experimental API for waveform iteration
+void readBlockVectorData(int dataID, int size, const int* valueIndices, double relativeReadTime, double* values) const;
+```
+
+`relativeReadTime` describes the time relatively to the beginning of the current time step. This means that `relativeReadTime = 0` gives us access to data at the beginning of the time step. Since we will call `advance(dt)` at a later point in time to finalize the time step, `relativeReadTime = dt` gives us access to data at the end of the time step.
+
+{% note %}
+The functionality of `writeBlockVectorData` remains unchanged, because the data at the beginning and at the end of the window are sufficient to create a linear interpolant over the window. Therefore, all samples but the one from the very last time step in the time window are ignored. This might, however, change in the future (see [precice/#1171](https://github.com/precice/precice/issues/1171)).
+{% endnote %}
+
+The experimental API has to be activated in the configuration file via the `experimental` attribute. This allows us to define the order of the interpolant in the `read-data` tag of the corresponding `participant`. The default is constant interpolation (`waveform-order="0"`). The following example uses `waveform-order="1"` and, therefore, linear interpolation:
+
+```xml
+<solver-interface experimental="true" ... >
+...
+    <participant name="FluidSolver">
+        <use-mesh name="FluidMesh" provide="yes"/>
+        <write-data name="Forces" mesh="MyMesh"/>
+        <read-data name="Displacements" mesh="FluidMesh" waveform-order="1"/>
+    </participant>
+...
+</solver-interface>
+```
+
+## Usage example
+
+We are now ready to extend the example from ["Step 6 - Implicit coupling"](couple-your-code-implicit-coupling.html) to use waveforms. Only few changes are necessary to sample the `Displacements` at the middle of the time window, which might be necessary for our specific application:
+
+```cpp
+// no relevant changes for initialization
+// See "Step 6 - Implicit coupling"
+...
+precice_dt = precice.initialize();
+while (not simulationDone()){ // time loop
+  // write checkpoint
+  ...
+  dt = beginTimeStep(); // e.g. compute adaptive dt 
+  dt = min(precice_dt, dt);
+  if (precice.isReadDataAvailable()){ // always true, because we can sample at arbitrary points
```

Comment:
No. I think `isReadDataAvailable` should always return `true`, even if `order=0` is configured. As soon as a user says "I want to use the experimental API", sampling is allowed by providing the additional argument to `readBlockScalarData` and therefore data is also available. Currently, the implementation is [as follows](https://github.com/BenjaminRodenberg/precice/blob/438de8f881fbb9b7c3bf93d612da96e2935bcc3f/src/precice/impl/SolverInterfaceImpl.cpp#L542-L550):

```
bool SolverInterfaceImpl::isReadDataAvailable() const
{
  PRECICE_TRACE();
  PRECICE_CHECK(_state != State::Constructed, "initialize() has to be called before isReadDataAvailable().");
  PRECICE_CHECK(_state != State::Finalized, "isReadDataAvailable() cannot be called after finalize().");
  bool available = _couplingScheme->hasDataBeenReceived();
  available |= (_allowsExperimental && _couplingScheme->hasInitialDataBeenReceived()); // if waveform relaxation is allowed, we always return true as soon as initialData is available.
  return available;
}
```

This is simpler and does not introduce another exception, because as soon as you are using the experimental API, `isReadDataAvailable` almost always returns true. With waveform relaxation I also think that the actual purpose of this API method becomes a bit unclear.

Reply:
But `experimental` includes many things, not only waveform relaxation.
The purpose of this function is to avoid unnecessary reads from the user. Imagine they have to do some costly conversion after calling read. In this sense, it should be false if `order=0`. It would be meaningless to read then, no?

Reply:
Let's add documentation about this later. The usage example works like it is right now.
I assume you changed the implementation in meanwhile? That one still needs fixing in my opinion.

Reply:
I added 68f0587cf4a389a1f01892d84f6b8804c5fba004. This reflects the current state as accurate as possible. I think it helps to say something (not too much) about the behavior of `isReadDataAvailable` here. We can add the detailed documentation as soon as https://github.com/precice/precice/issues/1223 is converged.